### PR TITLE
Sobol Method implementation

### PR DESCRIPTION
**Pull request by Vaibhavdixit02**

### CODE REVIEW

File: test/chemical_model.jl [hunk 1]
```diff
@@ -0,0 +1,20 @@
+using DiffEqSensitivity, OrdinaryDiffEq, ParameterizedFunctions
```
Comment (ChrisRackauckas) [inline test/chemical_model.jl]: This should go to diffeqproblemlibrary (with the analytical solution)

File: src/sobol_sensitivity.jl [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    f = function (p)
+        prob1 = remake(prob;p=p)
+        Array(solve(prob1,alg;saveat=t))
+    end
+
+    calc_mean_var(f,p_range,p_steps,N)
+end
+
+function calc_mean_var(f,p_range,p_steps,N)
+    p_values = [collect(linspace(p_range[i][1],p_range[i][2],p_steps[i])) for i in 1:length(p_range)]
```
Comment (ChrisRackauckas) [inline src/sobol_sensitivity.jl]: you shouldn't need to build a set of parameter values

File: src/sobol_sensitivity.jl [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    p = [rand(p_values[j]) for j in 1:length(p_values)]
```
Comment (ChrisRackauckas) [inline src/sobol_sensitivity.jl]: should be continuous variation

File: src/sobol_sensitivity.jl [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        y0 = [[0.0 for j in 1:size(y1)[1]] for i in 1:size(y1)[2]]
+        v = [[0.0 for j in 1:size(y1)[1]] for i in 1:size(y1)[2]]
+    else
+        y0 = [0.0 for j in 1:size(y1)[1]]
+        v = [0.0 for j in 1:size(y1)[1]]
+    end
+    for i in 1:N
+        p = [rand(p_values[j]) for j in 1:length(p_values)]
+        y1 = f(p)
+        if length(size(y1)) != 1
```
Comment (ChrisRackauckas) [inline src/sobol_sensitivity.jl]: why?

File: src/sobol_sensitivity.jl [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+                v[j] .+= (y1[j]).^2
+            end
+        else
+            y0 .+= y1
+            v .+= y1.^2
+        end
+    end
+    y0 = y0./N
+    y0_sq = [i.^2 for i in y0]
+    v = v./N .- y0_sq
```
Comment (ChrisRackauckas) [inline src/sobol_sensitivity.jl]: v's not right. You never do the second sample.

File: test/chemical_model.jl (hunk 1 above)
Comment (Vaibhavdixit02) [reply test/chemical_model.jl]: Will move it once the method is ready, currently using it as quick tests for implementation.

**Comment by Vaibhavdixit02** [comment]
@ChrisRackauckas ping.

File: src/sobol_sensitivity.jl [hunk 6]
```diff
@@ -0,0 +1,109 @@
+function give_rand_y(f,p_range,p_fixed=nothing,indices=nothing)
+    if p_fixed == nothing
+        p = [rand(p_range[j][1]:1e-6:p_range[j][2]) for j in 1:length(p_range)]
```
Comment (ChrisRackauckas) [inline src/sobol_sensitivity.jl]: why are these choices discrete?

File: src/sobol_sensitivity.jl [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    end
+    for i in 1:N
+        y1 = give_rand_y(f,p_range)
+        y0 .+= y1
+        y1 = give_rand_y(f,p_range)
+        v .+= y1.^2
+    end
+    y0 = y0./N
+    y0_sq = [i.^2 for i in y0]
+    v = v./N .- y0_sq
```
[the remark this answers is not in the document]
Comment (Vaibhavdixit02) [reply src/sobol_sensitivity.jl]: Yeah that is how it is defined, this calculation is mentioned explicitly in the thesis.

File: src/sobol_sensitivity.jl (hunk 6 above)
Comment (Vaibhavdixit02) [reply src/sobol_sensitivity.jl]: I expected `rand(p_range[j][1]:p_range[j][2])` to work as a continuous interval but it wasn't correct, what would be the syntax for that?

Comment (ChrisRackauckas) [reply src/sobol_sensitivity.jl]: `(p_range[j][2]-p_range[j][1])*rand() + p_range[j][1]`

File: src/sobol_sensitivity.jl (hunk 7 above)
[the remark this answers is not in the document]
Comment (ChrisRackauckas) [reply src/sobol_sensitivity.jl]: okay.

[the remark this answers is not in the document]
Comment (Vaibhavdixit02) [reply src/sobol_sensitivity.jl]: But this is giving significantly poor results (variations in the output in the order of `1e-1`) in my runs locally, should we change it?

[the remark this answers is not in the document]
Comment (Vaibhavdixit02) [reply src/sobol_sensitivity.jl]: With `N` being of the order `1e5`

**Comment by ChrisRackauckas** [comment]
Which paper are you working from?

**Comment by Vaibhavdixit02** [comment]
@ChrisRackauckas I primarily used the UCL thesis as reference.

File: src/sobol_sensitivity.jl [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+            else
+                p[i] = (p_range[i][2] -p_range[i][1])*rand() + p_range[i][1]
+            end
+        end
+    end
+    p
+end
+
+function calc_mean_var(f,p_range,N)
+    y1 = give_rand_y(f,p_range)
```
Comment (ChrisRackauckas) [inline src/sobol_sensitivity.jl]: why is this here?

Comment (ChrisRackauckas) [inline src/sobol_sensitivity.jl]: where is `give_rand_y` defined?

File: src/sobol_sensitivity.jl [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+        y0 = reshape(y0,size(y1)[1],size(y1)[2])
+        v = reshape(v,size(y1)[1],size(y1)[2])
+    end
+    for i in 1:N
+        y1 = give_rand_y(f,p_range)
+        y0 .+= y1
+        y1 = give_rand_y(f,p_range)
+        v .+= y1.^2
+    end
+    y0 = y0./N
```
Comment (ChrisRackauckas) [inline src/sobol_sensitivity.jl]: `@.` this whole thing.

File: src/sobol_sensitivity.jl [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+    y0_sq = [i.^2 for i in y0]
+    v = v./N .- y0_sq
```
Comment (ChrisRackauckas) [inline src/sobol_sensitivity.jl]: `@.` this whole thing.

File: src/sobol_sensitivity.jl [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+    y0 = [0.0 for j in 1:length(y1)]
+    v = [0.0 for j in 1:length(y1)]
+    if length(size(y1)) != 1
+        y0 = reshape(y0,size(y1)[1],size(y1)[2])
+        v = reshape(v,size(y1)[1],size(y1)[2])
+    end
+    for i in 1:N
+        y1 = give_rand_y(f,p_range)
+        y0 .+= y1
+        y1 = give_rand_y(f,p_range)
```
Comment (ChrisRackauckas) [inline src/sobol_sensitivity.jl]: why use a different one?

File: src/sobol_sensitivity.jl [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+    y0 = y0./N
+    y0_sq = [i.^2 for i in y0]
+    v = v./N .- y0_sq
+    y0,v
+end
+
+function first_order_var(f,p_range,N,y0)
+    ys = []
+    for i in 1:length(p_range)
+        y = [0.0 for i in 1:length(y0)]
```
Comment (ChrisRackauckas) [inline src/sobol_sensitivity.jl]: `zeros(y0)`?

File: src/sobol_sensitivity.jl [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+                p[i] = (p_range[i][2] -p_range[i][1])*rand() + p_range[i][1]
+            end
+        end
+    end
+    p
+end
+
+function calc_mean_var(f,p_range,N)
+    y1 = give_rand_y(f,p_range)
+    y0 = [0.0 for j in 1:length(y1)]
```
Comment (ChrisRackauckas) [inline src/sobol_sensitivity.jl]: `zeros(y1)`?

File: src/sobol_sensitivity.jl [hunk 14]
```diff
@@ -0,0 +1,144 @@
+function give_rand_p(f,p_range,p_fixed=nothing,indices=nothing)
+    if p_fixed == nothing
+        p = [(p_range[j][2] -p_range[j][1])*rand() + p_range[j][1] for j in 1:length(p_range)]
+    else
+        p = [0.0 for i in p_range]
```
Comment (ChrisRackauckas) [inline src/sobol_sensitivity.jl]: `zeros(p_range)`?

File: src/sobol_sensitivity.jl [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+    ys = []
+    for i in 1:length(p_range)
+        y = [0.0 for i in 1:length(y0)]
+        if length(size(y0)) != 1
+            y = reshape(y,size(y0)[1],size(y0)[2])
+        end
+        for j in 1:N
+            p2 = give_rand_p(f,p_range)
+            p1 = give_rand_p(f,p_range,[p2[i]],[i])
+            yer = Array(f(p1)) .* Array(f(p2))
```
Comment (ChrisRackauckas) [inline src/sobol_sensitivity.jl]: the mean/var calculations should utilize these same `f` calls. It should just be one function with some options to switch the sampling on/off for second order moments, but you can re-use a lot of computations here.

File: src/sobol_sensitivity.jl [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+            y .+= yer
+        end
+        y = y./N - y0.^2
+        push!(ys,copy(y))
+    end
+    ys
+end
+
+function sobol_sensitivity(f,p_range,N,order=0)
```
Comment (ChrisRackauckas) [inline src/sobol_sensitivity.jl]: default to second order.

Comment (ChrisRackauckas) [reply src/sobol_sensitivity.jl]: no. The total effects are far too expensive in most cases so they aren't really ever used. Sobol is useful over things like Morris since it can get the second order cases, so it's usually used because of its ability to get the second order interactions only. The total variance estimate then gives you a measure for how much the higher order interactions matter, which is usually low and justifies only doing the second order.

File: src/sobol_sensitivity.jl [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+    ys
+end
+
+function second_order_var(f,p_range,N,y0)
+    ys = []
+    for i in 1:length(p_range)
+        for j in i+1:length(p_range)
+            y = zeros(length(y0))
+            if length(size(y0)) != 1
+                y = reshape(y,size(y0)[1],size(y0)[2])
```
Comment (ChrisRackauckas) [inline src/sobol_sensitivity.jl]: why is this needed?

File: src/sobol_sensitivity.jl [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+            @. y += yer
+        end
+        y = @. y/N - y0^2
+        push!(ys,copy(y))
+    end
+    ys
+end
+
+function second_order_var(f,p_range,N,y0)
+    ys = []
```
Comment (ChrisRackauckas) [inline src/sobol_sensitivity.jl]: why not make this a matrix of the right size?

**Comment by ChrisRackauckas** [comment]
Just needs a few minor changes.

**Comment by ChrisRackauckas** [comment]
I'm merging since this is pretty much complete and improvements can happen in a further PR.

File: src/sobol_sensitivity.jl (hunk 18 above)
Comment (Vaibhavdixit02) [reply src/sobol_sensitivity.jl]: It was for the consistency across first order, second order and total order implementations, each would have to be initialised to be of different sizes that would require some calculation. Does this affect performance, I guess fixed size matrix will offer some sort of compiler optimisation, right?

File: src/sobol_sensitivity.jl (hunk 17 above)
Comment (Vaibhavdixit02) [reply src/sobol_sensitivity.jl]: For the solutions of DiffEqs?

File: src/sobol_sensitivity.jl (hunk 18 above)
Comment (ChrisRackauckas) [reply src/sobol_sensitivity.jl]: This effects performance in two ways. One, growing one-by-one is slower than making it once (though it's not so bad). But two, this is completely untyped so any access is not type inferred. If `f` is costly enough this doesn't matter though, but we might as well try.

File: src/sobol_sensitivity.jl (hunk 17 above)
Comment (ChrisRackauckas) [reply src/sobol_sensitivity.jl]: if you do `zeros(y0)` then you don't need to reshape. Why did you do `zeros` of the length?
